policy/release/repo_ids.rego

```diff
@@ -0,0 +1,170 @@
+# custom:
+#   short_name: rpm_repo_ids_valid
+#   failure_msg: 'RPM repo id check failed: %s'
+#   solution: >-
+#     Ensure every rpm is from a known and accepted repository and that the data in the SBOM
+#     correctly records that.
+#   # Todo: Until the sbom generation is upated, this will always fail,
+#   # so for now we don't include it in the redhat collection.
+#   # See https://issues.redhat.com/browse/STONEBLD-2638
+#   #collections:
+#   #- redhat
+#
+deny contains result if {
+	# Don't bother with this unless there's some data available
+	count(data_problems) == 0
+
+	some problem in repo_id_problems
+	result := lib.result_helper(rego.metadata.chain(), [problem])
+}
+
+data_problems contains reason if {
+	not known_repo_ids
+	reason := "The 'known_rpm_repositories' key was not found in the data"
+}
+
+data_problems contains reason if {
+	not is_array(known_repo_ids)
+	reason := "The 'known_rpm_repositories' data is not an array"
+}
+
+data_problems contains reason if {
+	count(known_repo_ids) == 0
+	reason := "The 'known_rpm_repositories' data is empty"
+}
+
+data_problems contains reason if {
+	is_array(known_repo_ids)
+	non_strings := [r |
+		some r in known_repo_ids
+		not is_string(r)
+	]
+	count(non_strings) > 0
+	reason := "The 'known_rpm_repositories' data is not a list of strings"
+}
+
+# See https://github.com/release-engineering/rhtap-ec-policy/tree/main/data
+# This list of repo_ids is not under the 'rule_data' key, so we don't use lib.rule_data.
+known_repo_ids := data.known_rpm_repositories
+
+repo_id_problems contains reason if {
+	bad_items := all_rpm_purls - raw_purls(all_parsed_rpm_purls)
+	count(bad_items) > 0
+
+	reason := sprintf(
+		"Some rpm components in the SBOM have purls that could not be parsed: %s",
+		[truncated_list_to_string(bad_items)],
+	)
+}
+
+repo_id_problems contains reason if {
+	bad_items := all_parsed_rpm_purls - all_parsed_purls_with_repo_ids
+	count(bad_items) > 0
+
+	reason := sprintf(
+		"Some rpm components in the SBOM did not specify a repository_id value in their purl: %s",
+		[truncated_list_to_string(raw_purls(bad_items))],
+	)
+}
+
+repo_id_problems contains reason if {
+	bad_items := all_parsed_purls_with_repo_ids - all_parsed_purls_with_known_repo_ids
+	count(bad_items) > 0
+
+	reason := sprintf(
+		"Some rpm components in the SBOM specify an unknown or disallowed repository_id: %s",
+		[truncated_list_to_string(raw_purls(bad_items))],
+	)
+}
+
+all_parsed_purls_with_known_repo_ids contains p if {
+	some p in all_parsed_purls_with_repo_ids
+	purl_repo_id(p.parsed) in known_repo_ids
+}
+
+all_parsed_purls_with_repo_ids contains p if {
+	some p in all_parsed_rpm_purls
+	purl_repo_id(p.parsed) # any repo id
+}
+
+# Keep the raw and parsed purl in a little struct-like
+all_parsed_rpm_purls contains p if {
+	some purl in all_rpm_purls
+	ec.purl.is_valid(purl)
+	p := {
+		"raw": purl,
+		"parsed": ec.purl.parse(purl),
+	}
```

Comment:
Perhaps, since from `parsed` only the `repository_id` is used? I think the `is_valid` can be skipped and we can rely on parsed not being nil, perhaps saves a bit of cpu time...

```suggestion
	parsed := ec.purl.parse(purl)
	parsed
	p := {
		"raw": purl,
		"repository_id": purl_repo_id(parsed),
	}
```

Reply:
The one difference is we spit out an `fmt.Errorf` when an attempted parse fails.

My reasoning is that I'd prefer not to produce an error in the output.

Not sure how important that is, but that's what I was thinking about.  ([This](https://github.com/enterprise-contract/ec-cli/pull/1923/commits/77732242cc8c533ae5c4163c0d0a3bbc1b8da2db) is related, since we used to also produce that error when calling is_valid.)

Reply:
Yeah, that makes sense, let's keep `ec.purl.is_valid`. Not sure if it's worth it, the second part is that we only the `repository_id` in the returned struct, perhaps a tiny bit of cpu/memory might be saved?

Reply:
Oh yeah good point.

Reply:
Fixed in latest revision.